tests/console_test.py

```diff
@@ -158,3 +160,117 @@ def test_interrupt(mocker: MockerFixture) -> None:
+            "--loglevel": "debug",
+            "--force": False,
+            "--reload": False,
+            "--description": "This is a description",
+            "--read-only": True,
+            "--uri": "testdb://test",
+            "--cost": 11.0,
+            "add-database": True,
+            "DATABASE": "testdb",
+            "REPOSITORY": None,
+        },
+    )
+    mocker.patch(
+        "datajunction.console.get_settings",
+        return_value=Settings(
+            index="sqlite:///dj.db",
+            repository=Path("/path/to/repository"),
+        ),
+    )
+
+    await console.main()
+    add_database.run.assert_called_with(
+        Path("/path/to/repository"),
+        database="testdb",
+        uri="testdb://test",
+        description="This is a description",
+        read_only=True,
+        cost=11.0,
+    )
+
+
+@pytest.mark.asyncio
+async def test_main_add_database_passing_repository(mocker: MockerFixture) -> None:
+    """
+    Test ``main`` with the "add-database" action.
+    """
+    add_database = mocker.patch("datajunction.console.add_database")
+    add_database.run = mocker.AsyncMock()
+
+    mocker.patch(
+        "datajunction.console.docopt",
+        return_value={
+            "--loglevel": "debug",
+            "--force": False,
+            "--reload": False,
+            "--description": "This is a description",
+            "--read-only": True,
+            "--uri": "testdb://test",
+            "--cost": 11.0,
+            "add-database": True,
+            "DATABASE": "testdb",
+            "REPOSITORY": "/path/to/another/repository",
+        },
+    )
+
+    await console.main()
+    add_database.run.assert_called_with(
+        Path("/path/to/another/repository"),
+        database="testdb",
+        uri="testdb://test",
+        description="This is a description",
+        read_only=True,
+        cost=11.0,
+    )
+
+
+@pytest.mark.asyncio
+async def test_main_add_database_raise_already_exists(
+    mocker: MockerFixture,
+    fs: FakeFilesystem,
+) -> None:
+    """
+    Test ``main`` with the "add-database" action raising when the database already exists
+    """
+    test_repo = "/foo"
+    fs.create_dir(test_repo)
+    mocker.patch(
+        "datajunction.console.docopt",
+        return_value={
+            "--loglevel": "debug",
+            "--force": False,
+            "--reload": False,
+            "--description": "This is a description",
+            "--read-only": True,
+            "--uri": "testdb://test",
+            "--cost": 11.0,
+            "add-database": True,
+            "DATABASE": "testdb",
+            "REPOSITORY": None,
+        },
+    )
+    mocker.patch(
+        "datajunction.console.get_settings",
+        return_value=Settings(
+            index="sqlite:///dj.db",
+            repository=Path(test_repo),
+        ),
+    )
+
+    await console.main()
+    await console.main()  # Run a second time to log an already exists exception
```

Comment:
One more thing, can you add a check like here:

https://github.com/DataJunction/datajunction/blob/e90897cf7b733aef682cee5d2a59b1e94c2fa714/tests/console_test.py#L102,L117

To test that `_logger.error` was called?

Reply:
Nice! I was a bit self-conscious about that one there and only implicitly knew it worked because of the coverage output. But what you linked is exactly what I was looking for! I updated the PR and also had to add an `__eq__` method to the `DJException` class [here](https://github.com/DataJunction/datajunction/pull/183/files#diff-63255a15e0e2e0c88eb0ddf47824d505ce575de93fac46086936047c3137d475R172) that checks that all attributes match (errors/warning lists, HTTP status code, etc).